usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -1101,19 +1139,20 @@ metaslab_block_find(avl_tree_t *t, uint64_t start, uint64_t size)
  * tree looking for a block that matches the specified criteria.
  */
 static uint64_t
-metaslab_block_picker(avl_tree_t *t, uint64_t *cursor, uint64_t size,
-    uint64_t align)
+metaslab_block_picker(metaslab_t *msp, avl_tree_t *t, uint64_t *cursor,
+    uint64_t size, uint64_t align)
 {
 	range_seg_t *rs = metaslab_block_find(t, *cursor, size);
 
-	while (rs != NULL) {
+	for (; rs != NULL; rs = AVL_NEXT(t, rs)) {
 		uint64_t offset = P2ROUNDUP(rs->rs_start, align);
 
-		if (offset + size <= rs->rs_end) {
+		if (offset + size <= rs->rs_end &&
+		    !metaslab_check_trim_conflict(msp, &offset, size, align,
+		    rs->rs_end)) {
```

Comment:
I'm not convinced any of these approaches will work for the following reasons:
1. If the storage is slow and there's lots to trim, then we can end up with all metaslabs doing trim at the same time. Yes, normally trims are staggered, but if the metaslabs take too long to trim, then we can conceivably end up with all metaslabs currently in a "trimming" state and your mostly empty pool still not being able to complete any allocations.
2. While possible, I worry about this bleeding through to an async spacemap update. Even autotrim executes as an async thread, and if metaslab decides to condense at the time we're currently auto-trimming it, we could conceivably write a space map with space seemingly allocated where it shouldn't be. If the box then crashes, this leaks space. Alternatively, I could write a whole new feature flag & format break the metaslab code to introduce an explicit "freeing in progress" object that space map would take into account when reconstructing the ms_tree at space_map_load. Seems to me to be a whole boatload of code just to save a couple of check lines here-and-there in metaslab.c.
3. Trims can take a long time to complete and halting spa_sync to wait for trim on a metaslab to complete could introduce huge and unpredictable latency bubbles, not to mention fundamentally break trim's original premise of being a "background low-priority operation".

As for metaslab_ff_alloc - should I just go ahead and kill the whole first-fit allocator as well?